umich_api/api_utils.py

```diff
@@ -114,21 +134,69 @@ def _api_call(self, api_call, access_token, method="GET", payload=None):
             return resp
 
     def get_next_page(self, response):
+        #type: (requests.Response) -> Dict
         if not 'next' in response.links:
             return None
         self.__log.debug(response.links)
         return parse_qs(urlparse(response.links['next']['url']).query)
 
     def get_access_token(self, token_url, client_scope):
-        # type: (str) -> str
-        """Retrieves an access token from the specified URL
-        
+        # type: (str, str) -> str
+        """Retrieves access token from the local tokens dictionary
+           If it's expired it will attempt to renew it
         :param token_url: Full URL to retrieve access token
         :type token_url: str
         :param client_scope: Client scope to retrieve this token for
         :type client_scope: str
+        :returns Access token str
+        :raises AttributeError If the oauth token could not be found
         """
+        token = self.get_oauth_token(token_url, client_scope)
+        if token is None:
+            raise AttributeError(f"Could not find oauth token for {token_url} {client_scope}")
+        return token.get("access_token")
 
+    def expire_token(self, token_url=None, client_scope=None):
+        """ Expires the token, given the token_url or the scope. For the scope it will expire all the tokens if given
+        
+        :param token_url: Token URL to expire
+        :type token_url: str
+        :param client_scope: Client scope to expire, will expire all tokens that use this 
+        :type client_scope: str 
+        :returns True if the token could be expired, False if there was some problem with the parameters or finding the token
+        """
+        # Try to get the tokenurl from the api for the client scope
+        if token_url is None:
+           token_url = self.scopes.get(client_scope, {}).get("token_url")
+
+        if token_url in self.tokens:
+            self.tokens[token_url]["expires_time"] = datetime.now()
+            self.__log.info(f"Token {token_url} has been expired")
+            return True
+        return False
+
+    def get_oauth_token(self, token_url, client_scope):
```

Comment:
Optional: This function looks too crowded, looks like lots going on and if possible chunking various part of the code to separate method might make it look simpler

Reply:
Yeah, I agree. I think having the tokens and scopes in their own class rather than directly accessed directly via a dict would improve the readability and decrease the complexity of this file. 

I don't think that's necessary to fix here but I can create a new issue about it. It was something I was thinking about while working through this. That's all internal and won't effect how this works. Good thing is with the test cases that should all be easy to change around.

Reply:
I created this issue as #19